python/ros_interface.py

```diff
@@ -0,0 +1,107 @@
+    def translate(self):
+        # Past values
+        self.past_gps_lat = self.gps_lat
+        self.past_gps_lon = self.gps_lon
+        self.past_gps_headingDegrees = self.gps_headingDegrees
+        self.past_gps_speedKmph = self.gps_speedKmph
+        self.past_wind_speedKmph = self.wind_speedKmph
+        self.past_wind_direction = self.wind_direction
+
+        # GPS sensors
+        self.gps_lat = self.checkErrAndAvg([self.past_gps_lat,
+                                            self.data.gps_0_latitude, self.data.gps_1_latitude])
+        self.gps_lon = self.checkErrAndAvg([self.past_gps_lon,
+                                            self.data.gps_0_longitude, self.data.gps_1_longitude])
+        self.gps_headingDegrees = self.checkErrAndAvg([self.past_gps_headingDegrees,
+                                                       self.convertDegrees(self.data.gps_0_true_heading),
+                                                       self.convertDegrees(self.data.gps_1_true_heading)])
+        self.gps_speedKmph = self.checkErrAndAvg([self.past_gps_speedKmph,
+                                                  self.data.gps_0_groundspeed * KNOTS_TO_KMPH,
+                                                  self.data.gps_1_groundspeed * KNOTS_TO_KMPH])
+        # Wind sensors
+        self.wind_speedKmph = self.checkErrAndAvg([self.past_wind_speedKmph,
+                                                   self.data.wind_sensor_0_speed * KNOTS_TO_KMPH,
+                                                   self.data.wind_sensor_1_speed * KNOTS_TO_KMPH,
+                                                   self.data.wind_sensor_2_speed * KNOTS_TO_KMPH])
+        self.wind_direction = self.checkErrAndAvg([self.past_wind_direction,
+                                                   self.data.wind_sensor_0_direction,
+                                                   self.data.wind_sensor_1_direction,
+                                                   self.data.wind_sensor_2_direction])
+
+    # (0 is north, 90 is east, etc - cw - degrees) -> (0 is east, 90 is north, etc - ccw - degrees)
+    def convertDegrees(self, degree):
+        converted = -1 * degree + 90
+        return converted if converted >= 0 else converted + 360
+
+    def checkErrAndAvg(self, arr):
+        err_free_arr = self.checkError(arr)
+        return float(sum(err_free_arr)) / len(err_free_arr)
+
+    # past value is first element of arr
+    # might causes errors: returns empty array, what if sensor readings aren't consistent
+    # could modify past value to be the average of the past few outputs (store in array)
+    def checkError(self, arr):
+        pastValue = arr[0]
+        presentValues = arr - arr[0]
+
+        errorFreeList = list()
+        for value in presentValues:
+            if abs(float((pastValue - value)) / pastValue) < pastValue * ERROR:
+                errorFreeList.append(value)
+
+        return numpy.array(errorFreeList)
```

Comment:
Is it necessary to check for errors, and not include sensor values that are outside a margin of error? Seems like it could cause some errors, which I mentioned in line 88

Reply:
Yeah I think what you have here is good for now, but it will definitely need to change later.